<?xml version="1.0"?>
<launch>

<!-- <arg name="meshfile" default="$(find uos_gazebo_worlds)/Media/models/avz_neu.dae" /> -->
<arg name="meshfile" default="/home/amock/hilti_uzh_tracking_area/reduced_mesh_09.ply" />
<arg name="device" default="cpu" />

<node pkg="rmcl" type="conv_pcl2_to_scan" name="conv_pcl2_to_scan" output="screen">
    <remap from="cloud" to="/ouster/points" />

    <param name="debug_cloud" type="bool" value="true" />
    <rosparam file="$(find rmcl)/config/sensors/OS0-64.yaml" />
    
    <!-- UNCOMMENT FOR REAL VELODYNE -->
    <!-- <rosparam file="$(find rmcl)/config/sensors/VLP-16.yaml" /> -->
</node>

<!-- ICP with given pose estimate -->

<group if="$(eval arg('device') == 'cpu')">

<node pkg="rmcl" type="ex_lidar_corrector_embree" name="lidar_corrector" 
    output="screen">
    <param name="map_file" type="string" value="$(arg meshfile)" />
    <param name="odom_frame" type="string" value="odom" />
    <param name="base_frame" type="string" value="base_link" />
    
    <param name="max_distance" type="double" value="0.5" />

    <param name="adaptive_max_dist" type="bool" value="true" />
    <param name="adaptive_max_dist_min" type="double" value="0.15" />

    <param name="poses" type="int" value="1" />
    <param name="tf_rate" type="double" value="50" />
    <param name="corr_rate_max" type="double" value="10" />



    <remap from="scan" to="/conv_pcl2_to_scan/scan"/>
    <!-- <remap from="pose" to="/move_base_simple/goal"/> -->
    <remap from="pose_wc" to="/initialpose" />
</node>

</group>


<group if="$(eval arg('device') == 'gpu')">

<node pkg="rmcl" type="lidar_corrector_optix" name="lidar_corrector" 
    output="screen">
    <param name="map_file" type="string" value="$(arg meshfile)" />
    <param name="odom_frame" type="string" value="odom" />
    <param name="base_frame" type="string" value="base_footprint" />
    
    <param name="max_distance" type="double" value="1.0" />
    
    <param name="adaptive_max_dist" type="bool" value="true" />
    <param name="adaptive_max_dist_min" type="double" value="0.15" />

    <param name="poses" type="int" value="1" />
    <param name="tf_rate" type="double" value="50" />
    <param name="corr_rate_max" type="double" value="600" />

    <remap from="scan" to="/pcl2_to_scan/scan"/>
    <!-- <remap from="pose" to="/move_base_simple/goal"/> -->
    <remap from="pose_wc" to="/initialpose" />
</node>

<!-- <node pkg="rmcl" type="segment_node_optix" name="segment_node" 
    output="screen">
    <param name="map_file" type="string" value="$(arg meshfile)" />
    <param name="map_frame" type="string" value="map" />

    <param name="max_distance" type="double" value="0.1" />
    <remap from="scan" to="/pcl2_to_scan/scan"/>

</node> -->

</group>




</launch>